### PR TITLE
Move settings into game state

### PR DESCRIPTION
**Pull request by rpherbig**
WIP, but this should serve as an example of what it would look like to move settings into game state.

The Processes would then focus on logic.

I picked `dance` at random as a first setting to move. Maybe I should focus on moving predicates first?

### CODE REVIEW

File: combat-trainer.lic [hunk 1]
```diff
@@ -55,12 +43,12 @@ class SetupProcess
       game_state.next_clean_up_step
       return true
     end
-    game_state.dancing = should_dance?
-    game_state.now_retreating = should_retreat?
+    game_state.update_dancing(DRRoom.npcs.length)
+    game_state.update_retreating(DRRoom.npcs.length)
```
Comment (rcuhljr) [inline combat-trainer.lic]: Could we just call game_state.update_npcs(DRRoom.npcs)? and have it deal with these?

File: combat-trainer.lic [hunk 2]
```diff
@@ -55,12 +43,12 @@ class SetupProcess
...
     if game_state.dancing
-      dance(game_state)
+      game_state.dance(@weapon_training)
```
Comment (rcuhljr) [inline combat-trainer.lic]: From reading this line I really don't have any good idea what this call would do.

File: combat-trainer.lic [hunk 3]
```diff
@@ -75,55 +63,29 @@ class SetupProcess
...
-    game_state.action_count >= @target_action_count || DRSkill.getxp(game_state.weapon_skill).to_i >= game_state.target_weapon_skill
-  end
-
-  def ending_ranged?(game_state)
-    game_state.was_retreating && !game_state.now_retreating
-  end
-
   def determine_next_to_train(game_state, weapon_training)
-    return unless skill_done?(game_state) || !weapon_training[game_state.weapon_skill] || ending_ranged?(game_state)
+    return unless game_state.skill_done?(DRSkill.getxp(game_state.weapon_skill).to_i) || !weapon_training[game_state.weapon_skill] || game_state.ending_ranged?
```
Comment (rcuhljr) [inline combat-trainer.lic]: Could we move `DRSkill.getxp(game_state.weapon_skill).to_i` into game_state?

File: combat-trainer.lic (hunk 1 above)
Comment (rpherbig) [inline combat-trainer.lic]: :+1:

File: combat-trainer.lic (hunk 3 above)
Comment (rpherbig) [inline combat-trainer.lic]: I considered that, but I kind of want to keep GameState "pure" in the sense of not knowing that level of detail. I thought that might also make it harder to test GameState. Maybe we pass in a proc?

File: combat-trainer.lic (hunk 2 above)
Comment (rpherbig) [inline combat-trainer.lic]: Thoughts on a better name? `select_dance_weapon`?

File: combat-trainer.lic (hunk 3 above)
Comment (rcuhljr) [inline combat-trainer.lic]: So it knows about experience values it just doesn't know how to get them?

Comment (rpherbig) [inline combat-trainer.lic]: It knows that there is a number it needs to compare, but not how that number is generated. Maybe that's a meaningless distinction? @kamarcum thoughts?

Will having GameState use DRSkill make that method any harder to test?

Comment (rcuhljr) [inline combat-trainer.lic]: No real testing difference, for all the testing we do.

Comment (rpherbig) [inline combat-trainer.lic]: :+1:

**Pull request merged**